File: src/main/scala/basic/tailRecursion.scala
package basic

/**
  * Created by ypokhrel on 5/31/2018.
  */

object App extends App{
	def normal_factorial(n:Int):Int = {
		if (n <= 0 ) 1
		else
			n * normal_factorial(n-1)
	}

	def tail_factorial(n: Int): Int = {
		def go(n:Int, acc:Int): Int = {
			if (n <=0) acc
			else go(n-1, n*acc)
		}

		go(n, 1)
	}

	println("factorial of 5 is : " + normal_factorial(5))
	println("factorial of 500000 is : " + tail_factorial(500000))
}

File: src/main/java/basic/tailRecursion.java
package basic;

/**
 * Created by ypokhrel on 5/31/2018.
 */
public class tailRecursion {
    public static void main(String[] args) {
        System.out.println("recurion of 5 is " + normal_factorial(5));
    }

    public static int normal_factorial(int n){
        if (n <= 1)
            return 1;
        else
            return n * normal_factorial(n - 1);
    }
}
